src/tracing/trace_flamegraph.c

```diff
@@ -0,0 +1,263 @@
+			/* We compare with 'memory' because 'prev_memory' is not memory when
+			   starting the function execution, 'memory' is. */
+			current_mem = zend_memory_usage(0);
+			if (current_mem < fse->memory) {
+				/* When memory is below 0, flamegraph generator will error, and
+				   you won't have a good visual.
+				   This happens when garbage collection happened during this
+				   function and freed something it didn't allocate, I guess. */
+				value = 0;
+			} else {
+				value = current_mem - fse->memory;
+			}
+			break;
+
+		case XDEBUG_TRACE_OPTION_FLAMEGRAPH_COST:
+			value = xdebug_get_nanotime() - fse->nanotime;
+			break;
+	}
+
+	return value;
+}
+
+xdebug_trace_flamegraph_context *xdebug_trace_flamegraph_init(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context;
+	char *used_fname;
+
+	tmp_flamegraph_context = xdmalloc(sizeof(xdebug_trace_flamegraph_context));
+	tmp_flamegraph_context->trace_file = xdebug_trace_open_file(fname, script_filename, options, (char**) &used_fname);
+	tmp_flamegraph_context->trace_filename = used_fname;
+
+	return tmp_flamegraph_context->trace_file ? tmp_flamegraph_context : NULL;
+}
+
+void *xdebug_trace_flamegraph_init_cost(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context = xdebug_trace_flamegraph_init(fname, script_filename, options);
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_COST;
+		tmp_flamegraph_context->stack = fg_stack_create();
+	}
+
+	return tmp_flamegraph_context;
+}
+
+void *xdebug_trace_flamegraph_init_mem(char *fname, zend_string *script_filename, long options)
+{
+	xdebug_trace_flamegraph_context *tmp_flamegraph_context = xdebug_trace_flamegraph_init(fname, script_filename, options);
+
+	if (tmp_flamegraph_context) {
+		tmp_flamegraph_context->mode = XDEBUG_TRACE_OPTION_FLAMEGRAPH_MEM;
+		tmp_flamegraph_context->stack = fg_stack_create();
+	}
+
+	return tmp_flamegraph_context;
+}
+
+void xdebug_trace_flamegraph_deinit(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	if (NULL != context->stack) {
+		fg_stack_free(context->stack);
+	}
+
+	fclose(context->trace_file);
+	context->trace_file = NULL;
+	xdfree(context->trace_filename);
+
+	xdfree(context);
+}
+
+void xdebug_trace_flamegraph_write_header(void *ctxt)
+{
+}
+
+void xdebug_trace_flamegraph_write_footer(void *ctxt)
+{
+}
+
+char *xdebug_trace_flamegraph_get_filename(void *ctxt)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+
+	return context->trace_filename;
+}
+
+void xdebug_trace_flamegraph_function_entry(void *ctxt, function_stack_entry *fse, int function_nr)
+{
+	xdebug_trace_flamegraph_context *context = (xdebug_trace_flamegraph_context*) ctxt;
+	flamegraph_stack_item *head = fg_stack_get_head(context->stack);
+	char *tmp_name, *prefix;
```

Comment:
I usually try to line up my variable names, albeit not consistently, such as at https://github.com/xdebug/xdebug/blob/master/src/debugger/handler_dbgp.c#L678-L680